File: week5/juyoung/problem3/problem3.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int N, L, R, X;
int A[15];		// 문제의 난이도
bool isSelected[15];
int answer = 0;

void dfs(int cnt) {
	if (cnt == N) {
		long long sum = 0;
		int max_value = -999999999;
		int min_value = 999999999;
		for (int i = 0; i < N; i++) {
			if (isSelected[i]) {
				sum += A[i];
				max_value = max(max_value, A[i]);
				min_value = min(min_value, A[i]);
			}
		}
		if (!(L <= sum && sum <= R)) return;
		if (max_value - min_value < X) return;
		answer++;
		return;
	}

	isSelected[cnt] = true;
	dfs(cnt + 1);
	isSelected[cnt] = false;
	dfs(cnt + 1);
}

int main() {
	cin >> N >> L >> R >> X;
	for (int i = 0; i < N; i++)
		cin >> A[i];
	dfs(0);
	cout << answer;
	return 0;
}

File: week5/sangmin/problem3/Problem3.java
package week5.sangmin.problem3;
import java.util.Scanner;

public class Problem3 {

    static int result = 0;
    static int N, X;
    static Long L, R;

    static int[] problem = new int[15];

    static void dfs(int beforeCount, int count, int idx, Long sum, Long max, Long min) {
        if(count >= 2) {
            if(L <= sum && sum <= R && X <= max - min && beforeCount != count) {
                result++;
            }
        }
        if(idx >= N || sum > R) {
            return;
        }
        // 포함 x
        dfs(count, count, idx + 1, sum, max, min);

        // 포함 o
        max = Math.max(max, problem[idx]);
        min = Math.min(min, problem[idx]);
        dfs(count, count + 1, idx + 1, sum + problem[idx], max, min);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        L = sc.nextLong();
        R = sc.nextLong();
        X = sc.nextInt();

        for(int i = 0; i < N; i++){
            problem[i] = sc.nextInt();
        }

        dfs(0, 0, 0, 0L, Long.MIN_VALUE, Long.MAX_VALUE);
        System.out.println(result);
    }
}